src/MSNMoney.py

```diff
@@ -1,46 +1,38 @@
-import logging
-from lxml import html
-
-def isfloat(value):
-  if value is None:
-    return False
-  try:
-    float(value)
-    return True
-  except ValueError:
-    return False
+import json
 
 class MSNMoney:
-  BASE_URL = 'https://www.msn.com/en-us/money/stockdetails/analysis?symbol={}'
-  PE_HIGH_KEY = 'P/E Ratio 5-Year High'
-  PE_LOW_KEY = 'P/E Ratio 5-Year Low'
-
-  @classmethod
-  def construct_url(cls, ticker_symbol,):
-    url = MSNMoney.BASE_URL.format(ticker_symbol)
-    return url
+  TICKER_URL = 'https://services.bingapis.com/contentservices-finance.csautosuggest/api/v1/Query?query={}&market=en-us'
+  BASE_URL = 'https://services.bingapis.com/contentservices-finance.financedataservice/api/v1/KeyRatios?stockId={}'
 
   def __init__(self, ticker_symbol):
     self.ticker_symbol = ticker_symbol.replace('.', '')
     self.pe_high = None
     self.pe_low = None
-    self.url = MSNMoney.construct_url(self.ticker_symbol)
 
-  def nextFloatFromIterator(self, iterator):
-    node = None
-    while node is None or not isfloat(node.text):
-      node = next(iterator)
-    return node.text
+  def get_ticker_autocomplete_url(self):
+    return self.TICKER_URL.format(self.ticker_symbol)
+
+  def get_financials_url(self, stock_id):
+    return self.BASE_URL.format(stock_id)
+
+  def extract_stock_id(self, content):
+    data = json.loads(content)
+    for ticker in data['data']['stocks']:
```

Comment:
To make these more fault-tolerant against URL change, can we make them in to `.get()` instead of using the `[]` syntax?

We can return a default empty dict in those cases.

`data.get('data', {}).get('stocks', [])`

--

Same with the other keys `RT00S` and `SecId`.

Reply:
Improved